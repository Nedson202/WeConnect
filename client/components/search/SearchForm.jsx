import React, { Component } from 'react';
import PropTypes from 'prop-types';
import '../../index.scss';

/**
 * @class SearchForm
 * 
 * @extends {Component}
 */
class SearchForm extends Component {
  /**
   * @description Creates an instance of search from.
   * 
   * @param {object} props 
   * 
   * @memberof Navbar
   */
  constructor(props) {
    super(props);

    this.state = {
      query: '',
      option: ''
    };

    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  /**
   * @description Handles input change
   * 
   * @param {any} event
   * 
   * @returns {undefined}
   * 
   * @memberof SearchForm
   */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value.toLowerCase() });
  }

  /**
   * @description Handles input submission
   * 
   * @param {any} event
   * 
   * @returns {undefined}
   * 
   * @memberof SearchForm
   */
  onSubmit(event) {
    const { query, option } = this.state;
    event.preventDefault();
    this.props.filterBusiness(option, query).then(() => {
      sessionStorage.setItem('query', JSON.stringify(this.state));
      this.context.router.history.push('/searchresult');
    });
  }

  /**
   * @description Renders the component to the dom
   * 
   * @returns {object} JSX object
   * 
   * @memberof BusinessProfile
   */
  render() {
    return (
      <div>
        <form className="form-inline my-2 my-lg-2" onSubmit={this.onSubmit}>
          <input
            className="form-control mr-sm-2"
            type="search"
            placeholder="Filter by location or category"
            name="query"
            onChange={this.onChange}
            value={this.state.query}
            required 
          />
          <select
            className="custom-select"
            name="option"
            onChange={this.onChange}
            value={this.state.option}
            required
          >
            <option value="">choose</option>
            <option value="name">Name</option>
            <option value="location">Location</option>
            <option value="category">Category</option>
          </select>
          <button className="btn btn-outline-success my-2 my-sm-0" type="submit"><i className="fa fa-search" /></button>
        </form>
      </div>
    );
  }
}

SearchForm.contextTypes = {
  router: PropTypes.object.isRequired
};

export default SearchForm;
