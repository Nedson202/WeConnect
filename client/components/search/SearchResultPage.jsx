import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import BusinessList from '../common/BusinessList';
import filterBusiness from '../../actions/businessQueryAction';
import '../../index.scss';

/**
 * @class SearcResultPage
 *
 * @extends {Component}
 */
class SearchResultPage extends Component {
  /**
   * @description set page title
   *
   * @returns {undefined}
   *
   * @memberof SearchResultPage
   */
  componentDidMount() {
    document.title = 'Search result';
  }

  /**
   * @description Renders the component to the dom
   *
   * @returns {object} JSX object
   *
   * @memberof SearchResultPage
   */
  render() {
    const { businesses, paginate } = this.props;
    const searchResult = businesses || [];
    const queryData = paginate ? paginate.queryData : {};

    const foundBusiness = (
      <div>
        { queryData !== undefined ? <h1 className="text-center textcase" id="business-found">Displaying business with {queryData.option}: {queryData.query}</h1> : null }
      </div>
    );

    return (
      <div>
        <div className="text-center business-profile" />
        {searchResult.length !== 0 ? foundBusiness : null}
        <BusinessList
          businesses={searchResult}
          filterBusiness={this.props.filterBusiness}
          paginate={paginate}
          queryData={queryData !== undefined ? queryData : {}}
        />
      </div>
    );
  }
}

SearchResultPage.propTypes = {
  filterBusiness: PropTypes.func.isRequired,
  businesses: PropTypes.array.isRequired,
  paginate: PropTypes.object.isRequired,
};

const mapStateToProps = (state) => {
  const { searchResult, paginationResult } = state.businesses;
  return {
    businesses: searchResult,
    paginate: paginationResult
  };
};

const mapDispatchToProps = dispatch => bindActionCreators({
  filterBusiness
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(SearchResultPage);
