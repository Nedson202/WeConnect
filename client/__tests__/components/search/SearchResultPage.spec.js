import React from 'react';
import { configure, shallow } from 'enzyme';
import thunk from 'redux-thunk';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import toJSON from 'enzyme-to-json';
import connectedSearchResultPage,  { SearchResultPage } 
from '../../../components/search/SearchResultPage';
import businessData from '../../__mockData__/business';
import { mapStateToProps } from '../../../components/search/SearchResultPage';

configure({ adapter: new Adapter() });

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const { allBusiness, paginationResult } = businessData;
let props, queryData;
const setup = () => {
  props = {
    businesses: allBusiness,
    paginate: paginationResult,
    filterBusiness: jest.fn(() => Promise.resolve()),
    onSubmit: jest.fn(() => Promise.resolve())
  };

  return shallow(<SearchResultPage {...props} />);
};


describe('Component: SearchResultPage', () => {
  let store;
  beforeEach(() => {
    global.document = {
      title: () => {}
    }
    global.queryData = {
      queryData: () => paginationResult.queryData
    }
  })

  it('should render', () => {
    const wrapper = setup();
    expect(wrapper.find('div').length).toBe(3);
    expect(wrapper.find('h1').length).toBe(1);
    expect(wrapper.find('BusinessList').length).toBe(1);
    wrapper.setProps({ businesses: [] });
    expect(wrapper.find('div').length).toBe(2);
    expect(toJSON(wrapper)).toMatchSnapshot();

    const state = {
      businesses: {
        searchResult: allBusiness,
        paginationResult
      }
    };

    expect(mapStateToProps(state).businesses).toEqual(allBusiness);
    const paginate = {
      page: 1,
      limit: 6,
      totalBusiness: 10,
      totalReviews: 10,
      queryData: {}
    }
    wrapper.setProps({ paginate });
    expect(wrapper.find('h1').length).toBe(0);
  }) 

  it('should render connect component', () => {
    const wrapper = shallow(<connectedSearchResultPage store={store} />);
    expect(wrapper.length).toBe(1);
  }) 
});