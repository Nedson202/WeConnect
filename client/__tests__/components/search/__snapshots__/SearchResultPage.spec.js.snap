// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: SearchResultPage should render 1`] = `
<div>
  <div
    className="text-center business-profile"
  />
  <BusinessList
    business={Object {}}
    businesses={Array []}
    fetchBusinesses={null}
    fetchBusinessesByUserId={null}
    filterBusiness={[MockFunction]}
    paginate={
      Object {
        "limit": 6,
        "page": 1,
        "queryData": Object {
          "option": "option",
        },
        "totalBusiness": 10,
        "totalReviews": 10,
      }
    }
    queryData={
      Object {
        "option": "option",
      }
    }
  />
</div>
`;
